docs/connector-development/tutorials/custom-python-connector/8-concurrency.md

```diff
@@ -0,0 +1,184 @@
+        ]
+        state_manager = ConnectorStateManager(stream_instance_map={s.name: s for s in synchronous_streams}, state=self._state)
+
+        configured_streams = []
+
+        for stream in synchronous_streams:
+
+            if stream.cursor_field:
+                cursor_field = CursorField(stream.cursor_field)
+                legacy_state = state_manager.get_stream_state(stream.name, stream.namespace)
+                cursor = ConcurrentCursor(
+                    stream.name,
+                    stream.namespace,
+                    legacy_state,
+                    self.message_repository,
+                    state_manager,
+                    stream.state_converter,
+                    cursor_field,
+                    self._get_slice_boundary_fields(stream, state_manager),
+                    _START_DATE,
+                )
+            else:
+                cursor = FinalStateCursor(stream.name, stream.namespace, self.message_repository)
+            configured_streams.append (
+                StreamFacade.create_from_stream(stream,
+                                                self,
+                                                _logger,
+                                                legacy_state,
+                                                cursor)
+                )
+        return configured_streams
+```
+The most interesting piece from this block is the use of `ConcurrentCursor` to support concurrent state management.
+
+The survey responses stream does not support incremental reads, so it's using a `FinalStateCursor` instead. The rest of the code change is mostly boilerplate.
+
+We'll also add a state converter to the `SurveyMonkeyBaseStream` to describe how the state cursor is formatted. We'll use the `EpochValueConcurrentStreamStateConverter` since the `get_updated_state` method returns the cursor as a timestamp
+```
+state_converter = EpochValueConcurrentStreamStateConverter()
+```
+
+Next we'll add a few missing constants:
+```
+_DEFAULT_CONCURRENCY = 10
+_MAX_CONCURRENCY = 10
+_RATE_LIMIT_PER_MINUTE = 120
+_logger = logging.getLogger("airbyte")
+```
+
+---
+:::info
+
+The substream isn't entirely concurrent because its stream_slices definition reads records from the parent stream concurrently:
+```python
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        for _slice in self._parent_stream.stream_slices():
+            for parent_record in self._parent_stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=_slice):
+                yield parent_record
+```
+
+This can be solved by implementing the connector using constructs from the concurrent CDK directly instead of wrapping synchronous streams in an adapter. This is left outside of the scope of this tutorial because no production connectors currently implement this.
```

Comment:
this is a very unfortunate broken window. We never got to productionizing support for concurrent substreams.